### PR TITLE
erlang 27.1 

### PR DESCRIPTION
**Pull request by paulo-ferraz-oliveira**
# Description

This updates the Erlang/OTP formula to 27.0.1.

## Further considerations

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

### CODE REVIEW

**Comment by paulo-ferraz-oliveira** [comment]
All commits squashed and force-pushed, as per the contributing guidelines.

**Comment by paulo-ferraz-oliveira** [comment]
@woodruffw, is [this](https://github.com/Homebrew/homebrew-core/actions/runs/10065808661/job/27825965676?pr=178020#step:4:45) something I should worry about?

**Comment by woodruffw** [comment]
> @woodruffw, is [this](https://github.com/Homebrew/homebrew-core/actions/runs/10065808661/job/27825965676?pr=178020#step:4:45) something I should worry about?

Hmm, I think those mean that those packages should be listed as dependencies to avoid accidental linkage breakages. But I'm not sure since I'm not as familiar with that part; CCing @Bo98 and @SMillerDev to confirm.

(The part that I'm not sure about is listing indirect deps -- I'm pretty sure we don't do that normally, but the warning about it presumably means something 🙂)

**Comment by paulo-ferraz-oliveira** [comment]
It's present on the [other pull request](https://github.com/Homebrew/homebrew-core/actions/runs/10066240477/job/27827270196?pr=178237#step:4:47), too, so I think this might've gone unnoticed (or is irrelevant), but in any case, with guidance I can fix it.

File: Formula/e/erlang.rb
Comment (chenrui333) [inline Formula/e/erlang.rb:89]: why removing this?

File: Formula/e/erlang.rb
Comment (chenrui333) [inline Formula/e/erlang.rb:41]:
```suggestion
  end

  on_linux do
    depends_on "mesa-glu"
  end

```

File: Formula/e/erlang.rb
Comment (chenrui333) [inline Formula/e/erlang.rb:37]:
```suggestion
  uses_from_macos "libxslt" => :build
  uses_from_macos "ncurses"
  uses_from_macos "zlib"
```

File: Formula/e/erlang.rb
Comment (paulo-ferraz-oliveira) [reply Formula/e/erlang.rb:89]: There's no more manpages, for 27.0.1. They might come back, later, but then the formula will have to be updated for it.

Comment (paulo-ferraz-oliveira) [reply Formula/e/erlang.rb:89]: My initial commit had more stuff in the body, but because I kept running into "Commit style" issues I was tweaking here and there, and eventually must've removed the body in desperation 😄

**Comment by ahankinson** [comment]
What needs to be done on this to bring it over the line? It would be good to get erlang updated in homebrew.

**Comment by paulo-ferraz-oliveira** [comment]
Was on vacation. Got back only recently. Haven't checked CI... I'm committed to moving this forward.

**Comment by ahankinson** [comment]
Just a quick look, and it seems that the Elixir formula has `Erlang 26` hard-coded in its test message, so when that changes, the build fails. 

https://github.com/Homebrew/homebrew-core/blob/master/Formula/e/elixir.rb#L34

**Comment by paulo-ferraz-oliveira** [comment]
Thanks for noticing, @ahankinson. I fixed and pushed. Let's see what happens now.

**Comment by paulo-ferraz-oliveira** [comment]
Failing because of:

* https://github.com/Homebrew/homebrew-core/actions/runs/10531415183/job/29184635908?pr=178020#step:3:155: don't understand the error (?)
* https://github.com/Homebrew/homebrew-core/actions/runs/10531415183/job/29184635908?pr=178020#step:3:373: seems it failed to fetch `rebar3_lint` (?)
* https://github.com/Homebrew/homebrew-core/actions/runs/10531415183/job/29184635908?pr=178020#step:3:448: something related to installation of LFE (?)
* https://github.com/Homebrew/homebrew-core/actions/runs/10531415183/job/29184635908?pr=178020#step:3:690: something related to wrangler (?)

A restart of CI might fix some of these, but not all, I guess.

**Comment by NelsonVides** [comment]
@paulo-ferraz-oliveira maybe we want to create a `ErlangAT26` formula as it exists for 25 already, for example. And then have all the dependents sign on the 26 version, until they themselves get updated? And probably even remove the support for 22.

@chenrui333 any thoughts on moving this forward?

**Comment by paulo-ferraz-oliveira** [comment]
Thanks, @NelsonVides, but I prefer to wait on feedback; I've already done updates in the past and gotten none; I also have a concurrent PR that might require attention.

**Comment by ahankinson** [comment]
I think the `erlang_ls` Formula needs to be pinned to `erlang@26`? 

They don't state that 27 is a supported version: https://github.com/erlang-ls/erlang_ls?tab=readme-ov-file#supported-otp-versions

**Comment by paulo-ferraz-oliveira** [comment]
I was convinced I had push changes, for the test results, as per my self-review, but it seems I'd stashed them and not pushed them. I did that now.

**Comment by paulo-ferraz-oliveira** [comment]
Updated the following dependants to "erlang@26" because they fail tests:
* emqx
* erlang_ls
* lfe
* wrangler

**Comment by NelsonVides** [comment]
I've just noticed that in the building args in the formula, there's still a `--enable-hipe` flag, now that's surely outdated! 😄 

Going through them a bit, `--enable-shared-zlib` is also outdated, as well as `--enable-smp-support` (OTP long ago dropped non-SMP to begin with) and `--enable-threads` and `--enable-wx`.

**Comment by paulo-ferraz-oliveira** [comment]
I did as little as possible to move forward and get it working (which I still can't because of CI 😢). We can add those improvements to removing OTP 22 from brew, too, since they'd be focused (HiPE wasn't in OTP 26 either, but at the time the person who updated it didn't care).

**Comment by paulo-ferraz-oliveira** [comment]
On the other hand, do you know how to add `erlang@26` and have the dependencies' tests use that? Because it seems erlang@26 is tested, passes tests, but then is uninstalled which means subsequent tests fail. Maybe it needs to be updated outside this PR first 🤷 (?)

**Comment by paulo-ferraz-oliveira** [comment]
I'm going back to the "two PRs to move this forward" process. One updates the failing dependencies to stay at OTP 26 (#183630). This one will do what it was doing before but without caring for updates to dependencies. I believe CI will initially fail but will start working after the previous one is merged and this one is rebased.

**Comment by paulo-ferraz-oliveira** [comment]
That didn't go well. I'm missing something here. `erlang@26` is seen as a new formula, but why wouldn't a new formula be tested as a dependency for existing formulas?

**Comment by paulo-ferraz-oliveira** [comment]
I'm gonna go with "add @26 as a new formula" and when that gets accepted I rebase this one over the main branch.

File: Formula/e/erlang.rb
Comment (chenrui333) [reply Formula/e/erlang.rb]: I see, make sense. 👍

File: Formula/e/erlang.rb
Comment (chenrui333) [reply Formula/e/erlang.rb:37]: this is more for the indirect linkage thing.

**Comment by chenrui333** [comment]
@paulo-ferraz-oliveira do you have confidence that we can ship this along with #183641 (as there are some dependency switches in here).

i also realize that lots erlang deps are build deps, which means we can ship after the formula update (it happens for rust formula build as well)

**Comment by paulo-ferraz-oliveira** [comment]
> we can ship this along with https://github.com/Homebrew/homebrew-core/pull/183641

We don't need to ship along, I guess. We first merge the other one, than I "fix" this one, if required, and we're done. We'll get a new formula...

**Comment by NelsonVides** [comment]
The other one looks good to me, looking forward to get it merged so this one can be continued 😄

**Comment by paulo-ferraz-oliveira** [comment]
This has, in time, accumulated conflicts, already. I'll update this branch, but it'd be nice if #183641 got some attention.

**Comment by paulo-ferraz-oliveira** [comment]
Rebased and force- pushed to resolve conflicts.

**Comment by ahankinson** [comment]
Maybe the draft status is holding it up?

**Comment by paulo-ferraz-oliveira** [comment]
It's not the draft. I need another PR merged before this one; the one that updates 26.x.

**Comment by NelsonVides** [comment]
OTP27.1 was just released and an automatic PR was created at https://github.com/Homebrew/homebrew-core/pull/191073, but it's also failing 🥲

**Comment by NelsonVides** [comment]
https://github.com/Homebrew/homebrew-core/pull/172171#issuecomment-2360940211 A note from @ferd on another issue :)

**Comment by paulo-ferraz-oliveira** [comment]
> but it's also failing

Most likely for the same reason it was failing before I attempted to update it with this PR: the compilation process changed.

@ferd's comment changes little but a caveat I'd removed. I can revisit this once the other pull request (#183641) gets merged.

**Comment by paulo-ferraz-oliveira** [comment]
If we can get this one merged auto-bumping to 27.1 shouldn't really cause issues. But if we keep waiting there's a change 27.1.1 comes out and this has to restart (I already had to rebase #183641 and fix merge conflicts before, because a different 26 came out).

**Comment by carlocab** [comment]
#183641 is merged. Let's try to get this PR done ASAP. Thanks!

**Comment by paulo-ferraz-oliveira** [comment]
Branch is rebased on top of main and pushed. As soon as CI is fixed (if any issues surface) I'll move it to Ready for review.

File: Formula/e/erlang.rb
Comment (carlocab) [inline Formula/e/erlang.rb:23]: Not sure where this diff came from but we shouldn't touch the `bottle` block. You may need to rebase your branch against `origin/master`.[^1]

[^1]: Rebase as in `git rebase`, please, and not `git merge`. (This may seem obvious but I've seen contributors use `git merge` when asked to rebase.)

File: Formula/e/erlang.rb
Comment (carlocab) [inline Formula/e/erlang.rb:87]: What is it downloading here? Can we avoid that with, e.g., `resource` blocks? We try not to have formulae download things during the build (and will block them from doing so at some point).

File: Formula/e/emqx.rb
Comment (carlocab) [inline Formula/e/emqx.rb:1]: You can split this out into a separate PR to simplify this one, since there is already an `erlang@26` formula.

File: Formula/e/erlang_ls.rb
Comment (carlocab) [inline Formula/e/erlang_ls.rb:1]: You can split this out into a separate PR to simplify this one, since there is already an `erlang@26` formula.

File: Formula/g/gleam.rb
Comment (carlocab) [inline Formula/g/gleam.rb:1]: You can split this out into a separate PR to simplify this one, since there is already an `erlang@26` formula.

File: Formula/l/lfe.rb
Comment (carlocab) [inline Formula/l/lfe.rb:1]: You can split this out into a separate PR to simplify this one, since there is already an `erlang@26` formula.

File: Formula/r/rebar3.rb
Comment (carlocab) [inline Formula/r/rebar3.rb:1]: You can split this out into a separate PR to simplify this one, since there is already an `erlang@26` formula.

File: Formula/w/wrangler.rb
Comment (carlocab) [inline Formula/w/wrangler.rb:1]: You can split this out into a separate PR to simplify this one, since there is already an `erlang@26` formula.

File: Formula/e/emqx.rb
Comment (paulo-ferraz-oliveira) [reply Formula/e/emqx.rb:1]: I could, but then I'd have to merge that one before this one, and again rebase, right? Or what do you mean?

File: Formula/e/erlang.rb
Comment (paulo-ferraz-oliveira) [reply Formula/e/erlang.rb:23]: I did rebase it, but in solving conflicts I chose my version of the content. What should I do here? Simply accept whatever's in the main branch?

Comment (paulo-ferraz-oliveira) [reply Formula/e/erlang.rb]: I've now rebased it again and removed the differences by squashing a new commit with that one.

File: Formula/e/erlang.rb
Comment (paulo-ferraz-oliveira) [reply Formula/e/erlang.rb:86]: This is what `download_ex_doc` does: https://github.com/erlang/otp/blob/57dbdfe82c1aa55ecf6e4a6f6f758d98e84abd24/otp_build#L1013.

If it's possible to do with a `resources` block I can do it, but otherwise this is Erlang/OTP specific, and may change with build scripts, so don't know how easy it'd be to maintain here.

**Comment by paulo-ferraz-oliveira** [comment]
Tests seem to be failing for lack of `escript` in the environment. `escript` is a binary created by Erlang/OTP's compilation. I'm not sure what's going on since `erlang@26` should create that.

File: Formula/e/emqx.rb
Comment (carlocab) [reply Formula/e/emqx.rb:1]: Yes, but each PR will be smaller and therefore run in CI faster.

File: Formula/e/erlang.rb
Comment (carlocab) [reply Formula/e/erlang.rb:86]: There's already a `resource` block in this formula. Below it, add,
```ruby
# https://github.com/erlang/otp/blob/#{version}/make/ex_doc_link
resource "ex_doc" do
  url "https://github.com/elixir-lang/ex_doc/releases/download/v0.34.1/ex_doc_otp_26"
  sha256 "d1e09ef6772132f36903fbb1c13d6972418b74ff2da71ab8e60fa3770fc56ec7"
end
```
and then in the `install` method, do
```
(buildpath/"bin").install resource("ex_doc") => "ex_doc"
```

File: Formula/e/emqx.rb
Comment (paulo-ferraz-oliveira) [reply Formula/e/emqx.rb:1]: Sure, I can do that. It's still failing for lack of `escript`, even with `=> :build`, and that I don't really understand...

File: Formula/w/wrangler.rb
Comment (paulo-ferraz-oliveira) [reply Formula/w/wrangler.rb:1]: #191438

File: Formula/e/emqx.rb
Comment (paulo-ferraz-oliveira) [reply Formula/e/emqx.rb:1]: #191439

File: Formula/l/lfe.rb
Comment (paulo-ferraz-oliveira) [reply Formula/l/lfe.rb:1]: #191440

File: Formula/r/rebar3.rb
Comment (paulo-ferraz-oliveira) [reply Formula/r/rebar3.rb:1]: #191441

File: Formula/g/gleam.rb
Comment (paulo-ferraz-oliveira) [reply Formula/g/gleam.rb:1]: #191441

File: Formula/e/erlang_ls.rb
Comment (paulo-ferraz-oliveira) [reply Formula/e/erlang_ls.rb:1]: #191441

**Comment by paulo-ferraz-oliveira** [comment]
I expect this one to start failing, now, until we get all other related pull requests in.

**Comment by paulo-ferraz-oliveira** [comment]
I resolved the conversation about creating new PRs because they're outdated and several PRs have already been opened, in the meantime. I'm gonna try and tackle the resources thing now.

File: Formula/e/erlang.rb
Comment (paulo-ferraz-oliveira) [reply Formula/e/erlang.rb:86]: Thanks for this. I'll copy-and-paste it and adapt locally, if required.

**Comment by paulo-ferraz-oliveira** [comment]
I've taken this change to update to 27.1 (all references including the doc mirror and `ex_doc` reference, as well as removing `automake` from the dependencies since it's not a requirement: "BTW why do you check for automake? Nothing in Erlang has ever used automake, only autoconf and make/gmake..." (from https://github.com/kerl/kerl/issues/546#issue-2488432542)

I'll now also review the build flags as some seem to have just been kept there but are no longer applicable (as @NelsonVides mentions above).

File: Formula/e/erlang.rb
Comment (carlocab) [inline Formula/e/erlang.rb:59]: At the top of the `install` method, let's do:
```ruby
ex_doc_url = (buildpath/"make/ex_doc_link").read.strip
odie "`ex_doc` resource needs updating!" if ex_doc_url != resource("ex_doc").url
```

File: Formula/e/erlang.rb
Comment (carlocab) [inline Formula/e/erlang.rb:54]:
```suggestion
  # https://github.com/erlang/otp/blob/#{version}/make/ex_doc_link
  resource "ex_doc" do
```
We should add this to make it a bit clearer how to update this resource.

File: Formula/e/erlang.rb
Comment (carlocab) [inline Formula/e/erlang.rb:90]: Oops, this snippet doesn't work. Let's do this:
```suggestion
    resource("ex_doc").stage do |r|
      (buildpath/"bin").install File.basename(r.url) => "ex_doc"
    end
```

**Comment by carlocab** [comment]
I'm going to apply my changes and then rebase to see what shape we're in after #191439, #191440, and #191441.

File: Formula/e/erlang.rb
Comment (carlocab) [inline Formula/e/erlang.rb:59]:
```suggestion
  def install
    ex_doc_url = (buildpath/"make/ex_doc_link").read.strip
    odie "`ex_doc` resource needs updating!" if ex_doc_url != resource("ex_doc").url
```

File: Formula/e/erlang.rb
Comment (carlocab) [inline Formula/e/erlang.rb:42]: Is this still needed after we've switched to downloading `ex_doc` using a `resource` block?

**Comment by carlocab** [comment]
Elixir is failing, but that seems to be due to a hard-coded assertion in the test.

Is this compatible with Erlang 27? Does it need to be rebuilt?

**Comment by NelsonVides** [comment]
> Elixir is failing, but that seems to be due to a hard-coded assertion in the test.
> 
> Is this compatible with Erlang 27? Does it need to be rebuilt?

https://github.com/Homebrew/homebrew-core/blob/a89ff871f0707e37c63d3e801442ee91a72d38e7/Formula/e/elixir.rb#L33 is hardcoded to be 26 there, probably that regex can be made more flexible.

Elixir 1.17 is indeed compatible with Erlang 27: https://hexdocs.pm/elixir/1.17.3/compatibility-and-deprecations.html#between-elixir-and-erlang-otp

**Comment by carlocab** [comment]
Thanks; I've opened #191460. I think we can ignore the `wrangler` failure; `wrangler` no longer seems to be maintained.

**Comment by carlocab** [comment]
Remaining failures should be fixed by #191459 and #191460.

**Review by carlocab — APPROVED** [review]

File: Formula/e/erlang.rb
Comment (paulo-ferraz-oliveira) [reply Formula/e/erlang.rb:42]: I don't think so; this is only the internal validation which Homebrew is probably already doing.

Comment (paulo-ferraz-oliveira) [reply Formula/e/erlang.rb]: I've since removed it, squashed the changes and force-pushed.

**Review by carlocab — APPROVED** [review]

**Comment by paulo-ferraz-oliveira** [comment]
This should be it. 🤞

**Comment by reneleonhardt** [comment]
@paulo-ferraz-oliveira amazing work to move the world forward to Erlang 26 and 27, thank you!
Is someone checking on RabbitMQ which announced since May that they are not supporting Erlang 27 yet 😅
https://www.rabbitmq.com/blog/2024/05/23/erlang27-support
https://www.rabbitmq.com/docs/which-erlang#erlang-27-support

> Warning: Erlang 27 (released in May 2024) is currently [not supported by any available RabbitMQ version](https://www.rabbitmq.com/blog/2024/05/23/erlang27-support). There are significant performance regressions when running RabbitMQ on Erlang 27.

So, can the dependency changed to the new alias `erlang@26` for the time being? 😄
https://github.com/Homebrew/homebrew-core/blob/master/Formula/r/rabbitmq.rb#L17

**Comment by carlocab** [comment]
> So, can the dependency be downgraded to the new alias `erlang@26` for the time being? 😄 [`master`/Formula/r/rabbitmq.rb#L17](https://github.com/Homebrew/homebrew-core/blob/master/Formula/r/rabbitmq.rb?rgh-link-date=2024-09-22T06%3A06%3A22Z#L17)

Pull requests welcome!

**Pull request merged**
